cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -91,11 +91,46 @@ struct TrueDiv {
 };
 
 struct FloorDiv {
-  template <typename T1, typename T2>
-  __device__ inline auto operator()(T1 const& lhs, T2 const& rhs)
-    -> decltype(floor(static_cast<double>(lhs) / static_cast<double>(rhs)))
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<(std::is_integral_v<std::common_type_t<TypeLhs, TypeRhs>> and
+                              std::is_signed_v<std::common_type_t<TypeLhs, TypeRhs>>)>* = nullptr>
+  __device__ inline auto operator()(TypeLhs x, TypeRhs y) -> decltype(x / y)
+  {
+    if ((x ^ y) >= 0) {
+      return x / y;
+    } else {
+      const auto quotient  = x / y;
+      const auto remainder = x % y;
+      return quotient - !!remainder;
```

Comment:
For my own education, what is going on here? :exploding_head:

Reply:
`!!` is a function $!! : \mathbb{N} \to {0, 1}$ with behaviour

$$
!! x = \begin{cases} 0 &\text{ if $ x = 0$} \\
1 & \text{ otherwise}
 \end{cases}
$$

If we have mixed sign arguments for floor division, if we have a non-zero remainder from the division we need to round towards negative infinity by subtracting 1 (this is a convenient branch-free way of getting 1).

Reply:
Nice trick. A comment in the code would be appropriate here.

Reply:
I've reworked this entire function to be branch-free and moved to less tricksy code (`remainder != 0`) which compiles to the same PTX.